File: src/books/archive-service.ts
/**
 * Book Archive Service
 * Handles archiving of books that are no longer accessible on the filesystem
 */

import type { Database } from "../../database.ts";
import * as bookData from "./data.ts";
import type { BookWithId } from "./book.ts";

export interface ArchiveResult {
    archivedCount: number;
    unarchivedCount: number;
    archivedBooks: BookWithId[];
    errors: Array<{ bookId: number; error: string }>;
}

export class ArchiveService {
    constructor(private database: Database) {}

    /**
     * Validate all active books and archive those whose files no longer exist
     */
    async validateAndArchiveMissingBooks(): Promise<ArchiveResult> {
        const result: ArchiveResult = {
            archivedCount: 0,
            unarchivedCount: 0,
            archivedBooks: [],
            errors: []
        };

        try {
            // Get all active books
            const activeBooks = await bookData.getActiveBooks(this.database);
            console.log(`📚 Archive Service: Validating ${activeBooks.length} active books`);

            for (const book of activeBooks) {
                try {
                    // Check if file exists
                    const fileExists = await this.checkFileExists(book.filepath);
                    
                    if (!fileExists) {
                        console.log(`📚 Archive Service: File not found, archiving book: ${book.name} (${book.filepath})`);
                        
                        // Archive the book with appropriate reason
                        await this.archiveBook(book.id, "File not found during validation scan");
                        
                        result.archivedCount++;
                        result.archivedBooks.push({
                            ...book,
                            status: 'archived',
                            archived_at: new Date().toISOString(),
                            archive_reason: "File not found during validation scan"
                        });
                    }
                } catch (error) {
                    console.error(`📚 Archive Service: Error validating book ${book.id}:`, error);
                    result.errors.push({
                        bookId: book.id,
                        error: error instanceof Error ? error.message : 'Unknown error'
                    });
                }
            }

            // Also check for previously archived books that might have been restored
            await this.checkForRestoredBooks(result);

            console.log(`📚 Archive Service: Validation complete. Archived: ${result.archivedCount}, Unarchived: ${result.unarchivedCount}, Errors: ${result.errors.length}`);
            
        } catch (error) {
            console.error('📚 Archive Service: Error during validation:', error);
            throw error;
        }

        return result;
    }

    /**
     * Check for previously archived books that might have been restored to the filesystem
     */
    private async checkForRestoredBooks(result: ArchiveResult): Promise<void> {
        try {
            const archivedBooks = await bookData.getArchivedBooks(this.database);
            
            for (const book of archivedBooks) {
                // Only check books archived due to missing files
                if (book.archive_reason?.includes("File not found") || book.archive_reason?.includes("missing")) {
                    try {
                        const fileExists = await this.checkFileExists(book.filepath);
                        
                        if (fileExists) {
                            console.log(`📚 Archive Service: File restored, unarchiving book: ${book.name} (${book.filepath})`);
                            await this.unarchiveBook(book.id);
                            result.unarchivedCount++;
                        }
                    } catch (error) {
                        console.error(`📚 Archive Service: Error checking restored book ${book.id}:`, error);
                        result.errors.push({
                            bookId: book.id,
                            error: error instanceof Error ? error.message : 'Unknown error'
                        });
                    }
                }
            }
        } catch (error) {
            console.error('📚 Archive Service: Error checking for restored books:', error);
        }
    }

    /**
     * Check if a file exists on the filesystem
     */
    private async checkFileExists(filepath: string): Promise<boolean> {
        try {
            const stat = await Deno.stat(filepath);
            return stat.isFile;
        } catch (error) {
            // If Deno.stat throws, the file doesn't exist or is inaccessible
            return false;
        }
    }

    /**
     * Archive a specific book
     */
    async archiveBook(bookId: number, reason: string): Promise<void> {
        await bookData.archiveBook(this.database, bookId, reason);
    }

    /**
     * Unarchive a specific book
     */
    async unarchiveBook(bookId: number): Promise<void> {
        await bookData.unarchiveBook(this.database, bookId);
    }

    /**
     * Get all archived books
     */
    async getArchivedBooks(): Promise<BookWithId[]> {
        return bookData.getArchivedBooks(this.database);
    }

    /**
     * Get all active books
     */
    async getActiveBooks(): Promise<BookWithId[]> {
        return bookData.getActiveBooks(this.database);
    }

    /**
     * Manually archive a book with a custom reason
     */
    async manualArchive(bookId: number, reason: string): Promise<void> {
        console.log(`📚 Archive Service: Manually archiving book ${bookId}: ${reason}`);
        await bookData.archiveBook(this.database, bookId, reason);
    }

    /**
     * Get archive statistics
     */
    async getArchiveStats(): Promise<{
        totalBooks: number;
        activeBooks: number;
        archivedBooks: number;
        archivedDueToMissingFiles: number;
    }> {
        const allBooks = await bookData.getAllBooks(this.database);
        const activeBooks = await bookData.getActiveBooks(this.database);
        const archivedBooks = await bookData.getArchivedBooks(this.database);
        
        const archivedDueToMissingFiles = archivedBooks.filter(book => 
            book.archive_reason?.includes("File not found") || 
            book.archive_reason?.includes("missing")
        ).length;

        return {
            totalBooks: allBooks.length,
            activeBooks: activeBooks.length,
            archivedBooks: archivedBooks.length,
            archivedDueToMissingFiles
        };
    }
}

File: src/books/archive_service.zig
const std = @import("std");
const sqlite = @import("sqlite");

pub const ArchiveService = struct {
    db: *sqlite.Db,
    allocator: std.mem.Allocator,
    
    pub fn init(allocator: std.mem.Allocator, db: *sqlite.Db) ArchiveService {
        return .{
            .allocator = allocator,
            .db = db,
        };
    }
    
    pub fn archiveBook(self: *ArchiveService, book_id: i64, reason: []const u8) !void {
        const query =
            \\UPDATE books 
            \\SET archived = 1,
            \\    archived_at = datetime('now'),
            \\    archive_reason = ?,
            \\    updated_at = datetime('now')
            \\WHERE id = ?
        ;
        
        var stmt = try self.db.prepare(query);
        defer stmt.deinit();
        
        stmt.reset();
        try stmt.bind(1, reason);
        try stmt.bind(2, book_id);
        
        try stmt.exec();
        std.log.info("📚 Archive Service: Archived book {d}: {s}", .{ book_id, reason });
    }
    
    pub fn unarchiveBook(self: *ArchiveService, book_id: i64) !void {
        const query =
            \\UPDATE books 
            \\SET archived = 0,
            \\    archived_at = NULL,
            \\    archive_reason = NULL,
            \\    updated_at = datetime('now')
            \\WHERE id = ?
        ;
        
        var stmt = try self.db.prepare(query);
        defer stmt.deinit();
        
        stmt.reset();
        try stmt.bind(1, book_id);
        
        try stmt.exec();
        std.log.info("📚 Archive Service: Unarchived book {d}", .{book_id});
    }
    
    fn checkFileExists(filepath: []const u8) bool {
        std.fs.cwd().access(filepath, .{}) catch {
            return false;
        };
        return true;
    }
    
    pub fn validateAndArchiveMissingBooks(self: *ArchiveService) !struct {
        archived_count: usize,
        unarchived_count: usize,
    } {
        var archived_count: usize = 0;
        var unarchived_count: usize = 0;
        
        // Get all active books
        {
            const query = "SELECT id, filepath, name FROM books WHERE archived = 0";
            var stmt = try self.db.prepare(query);
            defer stmt.deinit();
            
            stmt.reset();
            
            while (try stmt.step()) {
                const book_id = stmt.column(i64, 0);
                const filepath = stmt.column([]const u8, 1);
                const name = stmt.column([]const u8, 2);
                
                if (!checkFileExists(filepath)) {
                    std.log.info("📚 Archive Service: File not found, archiving book: {s} ({s})", .{ name, filepath });
                    try self.archiveBook(book_id, "File not found during validation scan");
                    archived_count += 1;
                }
            }
        }
        
        // Check for previously archived books that might have been restored
        {
            const query =
                \\SELECT id, filepath, name, archive_reason 
                \\FROM books 
                \\WHERE archived = 1 
                \\  AND (archive_reason LIKE '%File not found%' OR archive_reason LIKE '%missing%')
            ;
            var stmt = try self.db.prepare(query);
            defer stmt.deinit();
            
            stmt.reset();
            
            while (try stmt.step()) {
                const book_id = stmt.column(i64, 0);
                const filepath = stmt.column([]const u8, 1);
                const name = stmt.column([]const u8, 2);
                
                if (checkFileExists(filepath)) {
                    std.log.info("📚 Archive Service: File restored, unarchiving book: {s} ({s})", .{ name, filepath });
                    try self.unarchiveBook(book_id);
                    unarchived_count += 1;
                }
            }
        }
        
        std.log.info("📚 Archive Service: Validation complete. Archived: {d}, Unarchived: {d}", .{ archived_count, unarchived_count });
        
        return .{
            .archived_count = archived_count,
            .unarchived_count = unarchived_count,
        };
    }
    
    pub fn getArchiveStats(self: *ArchiveService) !struct {
        total_books: i64,
        active_books: i64,
        archived_books: i64,
        archived_due_to_missing_files: i64,
    } {
        const query =
            \\SELECT 
            \\  COUNT(*) as total_books,
            \\  SUM(CASE WHEN archived = 0 THEN 1 ELSE 0 END) as active_books,
            \\  SUM(CASE WHEN archived = 1 THEN 1 ELSE 0 END) as archived_books,
            \\  SUM(CASE WHEN archived = 1 AND (archive_reason LIKE '%File not found%' OR archive_reason LIKE '%missing%') THEN 1 ELSE 0 END) as archived_missing
            \\FROM books
        ;
        
        var stmt = try self.db.prepare(query);
        defer stmt.deinit();
        
        stmt.reset();
        
        if (try stmt.step()) {
            return .{
                .total_books = stmt.column(i64, 0),
                .active_books = stmt.column(i64, 1),
                .archived_books = stmt.column(i64, 2),
                .archived_due_to_missing_files = stmt.column(i64, 3),
            };
        }
        
        return .{
            .total_books = 0,
            .active_books = 0,
            .archived_books = 0,
            .archived_due_to_missing_files = 0,
        };
    }
};
